### PR TITLE
Merge sprint7Rf into integration

### PR DESCRIPTION
**Pull request by xaterz**
Several extra features and fixes have been implemented in this branch:
* **Added module search in the top navigation bar**
    * Allows the user to directly go to the Module Overview page of any module (without having to go to Module List and then click on the module)
    * Will autocomplete the module code/name based on what the user type. Autocomplete function is also used in the Prior Modules search form.
* **Added quota columns and number of students column to the Fixed Mounting and Tentative Mounting tables.** 
    * The purpose of adding these is for the user to compare/sort quotas and student demand across all modules. 
    * By default the columns are hidden, but user can tick the checkboxes above the table to show/hide these columns.
* **Fixed bug where quota cannot sort properly**
    * Problem is caused by the non-numerical quotas (e.g.  '-' and '?'). Now the quota columns can be sorted by treating the '-' and '?' as -2 and -1 respectively.
* **Added oversubscribed warning icon to warn user that module is oversubscribed**
    * When student demand > quota, will display orange icon beside student demand 
    * When module is not mounted but student demand > 0, will display red icon beside student demand 
    * Icons can be found in module overview, ay sem view, and fixed/tentative mounting tables
* **Added ay-sem selection for Overlapping Mods, Oversubscribed Mods, and Prior Mods Taken**
    * so that table has less data and can load faster
* **Added loading icon for all tables that have slow loading time**
    * Table will not be displayed until finish loading (does not affect nosetests)
    * While loading, will display a spinning loading icon
* **Reworked oversubscribed mod backend**
    * reduce the number of queries made to the database
    * Make the function able to distinguish '-' and '?' quotas (previously, the table is showing all the non-numerical quotas as '?', which is not correct)
* **Added modal popup for actions in module overview and aysem view**
    * Clicking on 'edit specific info', 'view overlap mods' or 'students taking module' will now display a pop-up prompting user to select an AY-Sem
    * The main reason for doing this is to make the effect of clicking on the action buttons consistent across the module overview page and the ay sem view page

Tagging all of you in this PR since a lot of these changes weren't discussed beforehand

### CODE REVIEW

File: components/handlers/fixed_module_mountings.py [hunk 1]
```diff
@@ -62,32 +66,58 @@ def populate_mounting_values(self):
...
+        self.populate_module_ay_sem_data()
         current_ay = model.get_current_ay()
 
         full_mounting_plan = self.full_mounting_plan
         # New modules will not be displayed in fixed mounting
-        full_mounting_plan = [subplan for subplan in full_mounting_plan
-                              if subplan[4].rstrip() == "Active"]
-
-        return RENDER.moduleMountingFixed(current_ay, full_mounting_plan)
+        full_mounting_plan = [subplan for subplan in full_mounting_plan
```
Comment (nlzz22) [inline components/handlers/fixed_module_mountings.py]: trailing whitespace.

File: components/handlers/module_taken_prior_to_others.py [hunk 2]
```diff
@@ -29,21 +29,24 @@ def GET(self, *test_data):
...
+            input_data = model.validate_input(web.input(), ["moduleAandB", "aysem"],
+                                              aysem_specific=False, attr_required=False)
 
+            target_ay_sem = None
             try:
-                module_A = input_data.moduleA
-                module_B = input_data.moduleB
                 target_ay_sem = input_data.aysem
+                module_A = input_data.moduleA
+                module_B = input_data.moduleB
```
Comment (nlzz22) [inline components/handlers/module_taken_prior_to_others.py]: trailing whitespace

File: components/handlers/module_taken_prior_to_others.py [hunk 3]
```diff
@@ -29,21 +29,24 @@ def GET(self, *test_data):
...
+                module_pairs = None
+                if target_ay_sem is not None:
+                    module_pairs = model.get_modA_taken_prior_to_modB(target_ay_sem)
 
                 # Get a list of all AY-Sems (for users to select)
                 all_ay_sems = model.get_all_ay_sems()
 
-                return RENDER.modulesTakenPriorToOthers(module_pairs, all_ay_sems, None, None,
-                                                        None, None, None, None)
+                return RENDER.modulesTakenPriorToOthers("all_pairs", module_pairs, all_ay_sems, None, None,
```
Comment (nlzz22) [inline components/handlers/module_taken_prior_to_others.py]: this line is too long

File: components/handlers/oversub_mod.py [hunk 4]
```diff
@@ -7,21 +7,111 @@
 from app import RENDER
```
Comment (nlzz22) [inline components/handlers/oversub_mod.py]: Missing pylint in this file

File: components/handlers/tentative_module_mountings.py [hunk 5]
```diff
@@ -19,7 +19,9 @@ class Tentative(object):
     def __init__(self):
         '''
             Full_mounting_plan is a list of 'subplans'
-            Each subplan is a list of 4 attributes (code, name, sem 1 mounting, sem 2 mounting)
+            Each subplan is a list of 8 attributes
```
Comment (nlzz22) [inline components/handlers/tentative_module_mountings.py]: trailing whitespace

File: components/handlers/tentative_module_mountings.py [hunk 6]
```diff
@@ -19,7 +19,9 @@ class Tentative(object):
     def __init__(self):
         '''
             Full_mounting_plan is a list of 'subplans'
-            Each subplan is a list of 4 attributes (code, name, sem 1 mounting, sem 2 mounting)
+            Each subplan is a list of 8 attributes 
+            (code, name, sem 1 mounting, sem 2 mounting, sem 1 quota, sem 2 quota,
+            number of students taking in sem 1, number of students taking in sem 2,)
```
Comment (nlzz22) [inline components/handlers/tentative_module_mountings.py]: additional comma character "," before the closing bracket.

File: components/helper.py [hunk 7]
```diff
@@ -13,7 +13,6 @@
     model = modules['components.model']
```
Comment (nlzz22) [inline components/helper.py]: please pylint this file

File: components/helper.py [hunk 8]
```diff
@@ -357,6 +356,8 @@ def validate_input(input_data, input_types, is_future=False,
         show_404:
             Set to False if don't want to return 404 page (function will return False instead)
     '''
+    from app import RENDER
```
Comment (nlzz22) [inline components/helper.py]: Shouldn't the import statement be at the top of the file ?

File: test/test_app.py [hunk 9]
```diff
@@ -177,3 +177,15 @@ def test_goto_mods_specific_size(self):
...
+    def test_goto_edit_all_mountings_and_quotas(self):
+        '''
+            Test if user can access edit all mountings and quotas
+            page without request errors
+        '''
+        root = self.test_app.get('/')
+
+        response = root.click(linkid="home-page", href="/editAll")
+
+        assert_equal(response.status, 200)
```
Comment (nlzz22) [inline test/test_app.py]: final newline missing

File: test/test_edit_all_mountings_and_quotas.py [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+    TABLE_HEADER_MOUNTED_IN_SEM1 = '<th>Mounted In Sem 1</th>'
+    TABLE_HEADER_SEM1_QUOTA = '<th>Sem 1 Quota</th>'
+    TABLE_HEADER_MOUNTED_IN_SEM2 = '<th>Mounted In Sem 2</th>'
+    TABLE_HEADER_SEM2_QUOTA = '<th>Sem 2 Quota</th>'
+    TABLE_HEADER_REMOVE = '<th>Remove</th>'
+    TABLE_NO_MODULES = 'No modules selected'
+
+    BUTTON_SAVE_CHANGES = '<button type="submit" class="btn btn-primary edit-all-bottom-button" ' +\
+                          'data-toggle="tooltip" data-placement="bottom" title="Save changes" '
+    BUTTON_RESET_TABLE = '<button type="button" id="edit-all-reset-button" class="btn btn-primary edit-all-bottom-button" ' +\
```
Comment (nlzz22) [inline test/test_edit_all_mountings_and_quotas.py]: line too long

File: test/test_edit_all_mountings_and_quotas.py [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+
+
+    def test_reset_edit_all(self):
+        '''
+            Tests that the Edit All table can be reset
+        '''
+        root = self.test_app.get(self.PAGE_URL)
+
+        response = root.clickbutton(buttonid='edit-all-reset-button')
+        response.mustcontain(self.TABLE_NO_MODULES)
```
Comment (nlzz22) [inline test/test_edit_all_mountings_and_quotas.py]: missing final newline

File: test/test_module_mounting_fixed.py [hunk 12]
```diff
@@ -46,10 +46,16 @@ class TestCode(object):
```
Comment (nlzz22) [inline test/test_module_mounting_fixed.py]: please pylint this file

File: test/test_module_mounting_tentative.py [hunk 13]
```diff
@@ -43,12 +43,18 @@ class TestCode(object):
                                  'data-toggle="tooltip" data-placement="bottom" ' +\
```
Comment (nlzz22) [inline test/test_module_mounting_tentative.py]: please pylint this file

File: test/test_non_overlap_module.py [hunk 14]
```diff
@@ -33,9 +33,9 @@ class TestCode(object):
            'data-placement="bottom" title="By default, these pairs are shown for the '+\
            'current AY-Semester." class="glyphicon glyphicon-info-sign" title="By '+\
            'default, these pairs are shown for the current AY-Semester."></span></p>'
-    FORM = '<form id="ay-form" class="form-inline" action="/nonOverlappingModules" method="post">'
-    SELECT_LABEL = '<label for="ay-sem">Select AY-Sem:</label>'
-    SELECT_ELEMENT = '<select class="form-control" name="aysem">'
+    FORM = '<form id="ay-form" class="form-inline aysem-dropdown" action="/nonOverlappingModules" method="post">'
```
Comment (nlzz22) [inline test/test_non_overlap_module.py]: line too long

File: test/test_overlapping_modules.py [hunk 15]
```diff
@@ -23,9 +23,11 @@ class TestCode(object):
     TABLE_COLUMN_MODULE_1_NAME = '<th>Name of Module 1</th>'
     TABLE_COLUMN_MODULE_2 = '<th>Module 2</th>'
     TABLE_COLUMN_MODULE_2_NAME = '<th>Name of Module 2</th>'
-    TABLE_COLUMN_AY_SEM = '<th>For AY-Sem</th>'
     TABLE_COLUMN_NUM_STUDENTS = '<th>Number of Students</th>'
 
+    FORM = '<form id="ay-form" class="form-inline aysem-dropdown" action="/overlappingModules" method="get">'
```
Comment (nlzz22) [inline test/test_overlapping_modules.py]: line is too long

File: test/test_overlapping_modules.py [hunk 16]
```diff
@@ -69,5 +71,7 @@ def test_page_contents(self):
         page.mustcontain(self.TABLE_COLUMN_MODULE_1_NAME)
         page.mustcontain(self.TABLE_COLUMN_MODULE_2)
         page.mustcontain(self.TABLE_COLUMN_MODULE_2_NAME)
-        page.mustcontain(self.TABLE_COLUMN_AY_SEM)
         page.mustcontain(self.TABLE_COLUMN_NUM_STUDENTS)
+        page.mustcontain(self.FORM)
+        page.mustcontain(self.SELECT_LABEL)
+        page.mustcontain(self.SELECT_ELEMENT)
```
Comment (nlzz22) [inline test/test_overlapping_modules.py]: final newline missing

File: test/test_students_affected_by_module.py [hunk 17]
```diff
@@ -14,9 +14,9 @@ class TestCode(object):
     '''
 
     URL_NORMAL = '/studentsAffectedByModule?code=IT1005&aysem=AY+17%2F18+Sem+1'
-    CONTENT_TITLE = '<h1 class="text-center">Students Planning To Take This Module</h1>'
-    CONTENT_DESCRIPTION = '<p class="text-center">Students who are planning'+\
-                          ' to take this module.</p>'
+    CONTENT_TITLE = '<h1 class="text-center">Students Taking This Module</h1>'
+    CONTENT_DESCRIPTION = '<p class="text-center">Show students who <b>have taken</b>, '+\
+                          'are <b>currently taking</b>, or are <b>planning to take</b> this module.</p>'
```
Comment (nlzz22) [inline test/test_students_affected_by_module.py]: this line is too long

File: test/test_view_individual_module.py [hunk 18]
```diff
@@ -49,31 +49,23 @@ class TestCode(object):
                                      ' name="edit-mounting-button" '+\
```
Comment (nlzz22) [inline test/test_view_individual_module.py]: please pylint this file

File: test/test_view_module.py [hunk 19]
```diff
@@ -51,25 +51,27 @@ class TestCode(object):
     CONTENT_INFO_TENTA = "Information for Future AYs (Tentative)"
```
Comment (nlzz22) [inline test/test_view_module.py]: please pylint this file

**Comment by nlzz22** [comment]
image

There is a bug with the Tentative Module Mounting sorting. (note: there is NO such problem in fixed mounting table) The sorting treats 4 > 37 >3 > 25, etc. (This is a problem for "number of students taking in sem 1", Sem 2 quota" and "number of students taking in sem 2") NOTE: there is NO such problem in "Sem 1 quota" though.

My guess is there is some missing variables here: 
image

which should target columns 3,4,6,7 instead of 3,5.

**Comment by a0129998** [comment]
image
suggestion:

move the magnifying glass outside of the search bar, and use `glyphicon glyphicon-search` for a more pleasant look.

**Comment by a0129998** [comment]
image
not sure if this is the correct implementation, if it is please change it to `demand >= supply`

**Comment by a0129998** [comment]
Bug:

image
`edit specific module info` is buggy and nolonger works. It is functional in `integraion`

**Comment by xaterz** [comment]
Fixed pylint issues, the tentative mounting table bug and the magnifying glass icon. @nlzz22 @a0129998 please review again.

Could not replicate the problem raised by @a0129998 regarding the action buttons in module overview not working. Unless the problem can be replicated with certainty, the buttons shall be assumed to be working fine.

**Pull request merged**